zkml/docs/src/llms-layers/qkv.md

```diff
@@ -0,0 +1,95 @@
+# QKV
+
+## Description of the Layer
+The **QKV** layer is the first layer found in each encoder/decoder of an LLM architecture, and it is employed to produce, for each word (or better, *token*) being processed by the LLM three vectors:
+
+- A query vector `q`
+- A key vector `k`
+- A value vector `v`
+
+These vectors are all computed from a vector `x` representing a token processed by the LLM. More specifically, each of this vector is obtained by multiplying `x` with corresponding weight matrices  $W_q$, $W_k$ and $W_v$, and then each resulting vector is added to a corresponding bias vector $B_q$, $B_k$ and $B_v$. Since an LLM usually needs to process more than 1 token at the same time, the computation of `q`, `k` and `v` vectors for all the input tokens is done in a "batched" fashon over matrices rather than over single vectors:
+
+- The `s` vectors $x_i \in \mathbb{R}^m$, each representing one of the `s` input tokens being processed, are concatenated in a matrix $X \in \mathbb{R}^{s \times e}$, where the $i$-th row of $X$ is equal to $X_i$
+- Define the unitary vector as $\mathbb{1}_s = 1^s$. Given the bias vectors $B_q \in \mathbb{R}^h$, $B_k \in \mathbb{R}^h$ and $B_v \in \mathbb{R}^h$, define the *bias matrices* $B_q^s \in \mathbb{R}^{s \times h}= \mathbb{1}_s*B_q^T$, $B_k^s \in \mathbb{R}^{s \times h}= \mathbb{1}_s*B_k^T$ and $B_v^s \in \mathbb{R}^{s \times h}= \mathbb{1}_s*B_v^T$; in other words, each of the $s$ rows in a bias matrix is equal to the corresponding bias vector 
+- Given the weight matrices $W_q \in \mathbb{R}^{e \times h}$, $W_k \in \mathbb{R}^{e \times h}$ and $W_v \in \mathbb{R}^{e \times h}$, the output matrices $Q = XW_q + B_q^s$, $K = XW_k + B_k^s$, $V = XW_v + B_v^s$ are computed; the $i$-th row of these matrices corresponds, respectively, to the query, key and value vectors of the $i$-th input token 
+
+The QKV layer is basically computing these three matrices `Q`, `K` and `V` from the matrix `X` representing the input tokens
+
+## Proving the Layer
+Proving the three matrix multiplications to compute `Q`, `K` and `V` can be done with a single *batched* sum-check. 
+
+The prover starts from 3 claims $y_Q$, $y_K,$ and $y_V$ about the MLE of output matrices `Q`, `K` and `V`, respectively. More specifically, the 3 claims are computed at the random points $r_Q = (r'_Q \in \mathbb{F}^{\log(s)}, r''_Q \in \mathbb{F}^{\log(h)})$, $r_K = (r'_K \in \mathbb{F}^{\log(s)}, r''_K \in \mathbb{F}^{\log(h)})$, $r_V = (r'_V \in \mathbb{F}^{\log(s)}, r''_V \in \mathbb{F}^{\log(h)})$. Furthermore, the prover has computed the MLEs of the weight matrices $W_q$, $W_k$ and $W_v$ and the MLEs of the bias vectors $B_q$, $B_k$ and $B_v$, which were commited to in a setup phase. 
```

Comment:
Can you make space in the paragrpahs, like use bullet points to describes the three points for example. ?

Reply:
Done in commit https://github.com/Lagrange-Labs/deep-prove/pull/123/commits/a8f86fba6856eed02b5775488b89e187fde4c33b